DirectProgramming/DPC++FPGA/ReferenceDesigns/snappy/src/main.cpp

```diff
@@ -0,0 +1,468 @@
+      fpga_tools::RoundUpToMultiple(in_count, kLiteralsPerCycle);
+
+  // read the expected output size from the start of the file
+  // this is used to size the output buffer
+  unsigned out_count = 0;
+  unsigned byte_idx = 0;
+  unsigned shift = 0;
+  bool keep_reading_preamble = true;
+  while (keep_reading_preamble) {
+    if (byte_idx > 4) {
+      std::cerr << "ERROR: uncompressed length should not span more than 5"
+                << " bytes\n";
+      std::terminate();
+    }
+    auto b = in_bytes[byte_idx];
+    keep_reading_preamble = (b >> 7) & 0x1;
+    out_count |= (b & 0x7F) << shift;
+    shift += 7;
+    byte_idx += 1;
+  }
+
+  std::vector<unsigned char> out_bytes(out_count);
+
+  // round up the output count to the nearest multiple of kLiteralsPerCycle
+  // this allows to ignore predicating the last writes to the output
+  int out_count_padded =
+      fpga_tools::RoundUpToMultiple(out_count, kLiteralsPerCycle);
+
+  // host variables for output from device
+  int decompressed_count_host;
+  unsigned preamble_count_host;
+
+  // track timing information in ms
+  std::vector<double> time(runs);
+
+  // input and output data pointers on the device using USM device allocations
+  unsigned char *in, *out;
+  unsigned* decompressed_count;
+  unsigned* preamble_count;
+
+  try {
+    // allocate memory on the device for the input and output
+    if ((in = malloc_device<unsigned char>(in_count_padded, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'in'\n";
+      std::terminate();
+    }
+    if ((out = malloc_device<unsigned char>(out_count_padded, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'out'\n";
+      std::terminate();
+    }
+    if ((decompressed_count = malloc_device<unsigned>(1, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'decompressed_count'\n";
+      std::terminate();
+    }
+    if ((preamble_count = malloc_device<unsigned>(1, q)) == nullptr) {
+      std::cerr << "ERROR: could not allocate space for 'preamble_count'\n";
+      std::terminate();
+    }
+
+    // copy the input data to the device memory and wait for the copy to finish
+    q.memcpy(in, in_bytes.data(), in_count * sizeof(unsigned char)).wait();
+
+    // run the design multiple times to increase the accuracy of the timing
+    for (int i = 0; i < runs; i++) {
+      std::cout << "Launching kernels for run " << i << std::endl;
+
+      // run the producer and consumer kernels
+      auto producer_event = SubmitProducer(q, in_count_padded, in);
+      auto consumer_event =
+          SubmitConsumer(q, out_count_padded, out, decompressed_count);
+
+      // run the decompression kernels
+      auto snappy_decompress_events =
+          SubmitSnappyDecompressKernels<InPipe, OutPipe, kLiteralsPerCycle>(
+              q, in_count, preamble_count);
+
+      // wait for the producer and consumer to finish
+      auto start = high_resolution_clock::now();
```

Comment:
Yes, I believe so.
The idea here is the following: in general, you will have these kernels running forever, constantly decoding the snappy input. So I want to time the producers/consumers. Once all the data has been consumed, the job is done (this is actually pessimistic because it include the time to teardown the producer/consumer kernel, but we consider that negligible with large enough data sets).